Traffic Lighters_version 2

Vlad-Slyzkoukhyi opened the pull request:
Create a dictionary for objetcs like traffic lighters.
Optimized the code in the TrafficLighterModes class.
Separated all methods what invoke different traffic lighters states.

Vlad-Slyzkoukhyi committed:
Add files via upload

Vlad-Slyzkoukhyi committed:
Add files via upload

Vlad-Slyzkoukhyi committed:
Merge branch 'Traffic_Lighters_version2' of https://github.com/Vlad-Slyzkoukhyi/Traffic-lighters into Traffic_Lighters_version2

Vlad-Slyzkoukhyi committed:
Merge branch 'Traffic_Lighters_version2' of https://github.com/Vlad-Slyzkoukhyi/Traffic-lighters into Traffic_Lighters_version2

Vlad-Slyzkoukhyi committed:
Merge branch 'Traffic_Lighters_version2' of https://github.com/Vlad-Slyzkoukhyi/Traffic-lighters into Traffic_Lighters_version2

ValeriiB commented on TrafficLighters_simple.jpg:
Let's think what is possible implementation of a read traffic light system. I believe that there is a a module that controls of all traffic lights in that crosroad. I would call it a TrafficLightsController or CrossRoadController. In you application you named it like TrafficLighterModes. That controller controlles states of the crossroad. How does it works: 
DayMode:
DO
  GET next state (circularly) from states in the day mode (you defined 10 states on you diagram)
  SET crossroad state
  WAIT
WHILE(DayMode)
Very simple, no?

Each state of a crossroad defines states of each road in that crossroad. Those roads are for MainRoad, SecondaryRoad, TramRoad and PedestrianRoad.

Can we describe those 10 states of the crossroad in a table (it may be some collection like a dictionary or a list or an array) where each record in the table represents a state (columns on you diagram) and each crossroad state has a collection of road states. 
Note, the table contains states and not TrafficLights
Road          |     1      |           2           |       3               | 4             ...
Main          | GREEN | BLINK GREEN | YELLOW          | REG        ...
Secondary | RED      | RED                | YELLOW RED   | GREEN   ...
Pedestrian | RED      | RED                | RED                 | RED        ...
Tram          | STOP    | STOP              | STOP               | STOP       ...

Then those states are represented by lamps in the traffic light. Does the controller work differently if let's say MainRoad has 1 or 2 or 500 traffic light presentation? No. So why the controller should manage MainRoad#1 and MainRoad#2? It should manage the state of MainRoad and the presentation of that MainRoad state it not its responsibility.

The Controller can now get a state from the table and set a stat for the crossroad where we set states for each road on the crossroad.
Now I can easily add/remove a new road to the system or add/remove new state to the crossroad. The logic of the controller will not change.
I even can define that table in some configuration file and I do not need to change the program.

Your solution is hardcoded and I cannot easily change the configuration of tthe system without changing the program.

Change your program accordingly.
For now you can hardcode the table of the states. But have in mind that that table contains the configuration of the system and to make it flaxible in the future we may want to define that table in a text file, so the program will read it on start.

ValeriiB commented:
Do not create new version every time. We have not finished any of them.

Vlad-Slyzkoukhyi committed:
New implementation with states

Vlad-Slyzkoukhyi committed:
Re-worked implementation with usind set of data for traffic lighters states and time.
Presentation menu was fully moved to presentation side.
Totally optimazed all clases.

ValeriiB commented on CrossRoadController.cs:
```diff
@@ -0,0 +1,374 @@
+        internal bool RightLamp { get; set; }
+        internal bool LeftLamp { get; set; }
+        internal bool MiddleLamp { get; set; }
+        internal bool BottomLamp { get; set; }
```
Controller should control traffic lights but not lamps.

ValeriiB commented on CrossRoadController.cs:
```diff
@@ -0,0 +1,374 @@
+        internal Dictionary<string, List<string>> DayModeStatesCollection = new()
+        {
+            {"States",    new(){"1","2","3","4","5", "6", "7", "8", "9", "10"} },
+            {"Main",      new(){"RED","REDYELLOW","GREEN","BLINKGREEN","YELLOW",   "RED",  "RED",       "RED",   "RED","RED"}},
```
Define states of the traffic lights as enum and not as a string.

ValeriiB commented on CrossRoadController.cs:
```diff
@@ -0,0 +1,374 @@
+            CrossRoadController crossRoadController = new();
+            do
+            {
+                time = TimeOfDay.TimeNow();
```
Do not tie the controller to a time of day. There should be a possibility to switch states by timer, by buttons on the UI, by any code in unit tests etc.

ValeriiB commented on CrossRoadController.cs:
```diff
@@ -0,0 +1,374 @@
+        internal const int mainTimer = 5000;
+        internal const int blinkTimer = 500;
+        internal const int numberOfChanges = 5;
+        internal const int yellowTimer = 2000;
```
If you think of the state macine, you will see that you change the state of the state machine and then the state machine stay in that state some repiod of time. Enstead of defining the times for blinking green, yellow atc, you can define in the table of states the time the state machine should be in that state. The your logic will be much much simpler: You get the state from the table, you gate time for the state from theat table then switch the state and wait the time.

ValeriiB commented on CrossRoadController.cs:
```diff
@@ -0,0 +1,374 @@
+                            Main = DayModeStatesCollection["Main"][i];
+                            Secondary = DayModeStatesCollection["Secondary"][i];
+                            Tram = DayModeStatesCollection["Tram"][i];
+                            Pedestrian = DayModeStatesCollection["Pedestrian"][i];
```
Think of iterating through the traffic lights in a cycle:
```
foreach(var stateName in DayModeStates)
{
    foreach(var trafficLightName in TrafficLightNames)
    { 
        var trafficLightState = TrafficLightStates[stateName][trafficLightName];
        var stateTime = TrafficLightStates[stateName]["Time"];
        trafficLights[trafficLightName].SetState(trafficLightState);
        Task.Delay(stateTime);
    }
}
```

ValeriiB commented on CrossRoadController.cs:
```diff
@@ -0,0 +1,374 @@
+                                await Task.Delay(blinkTimer);
+                                cycleTime += blinkTimer;
+                                crossRoadController.CrossRoadControllerEvent?.Invoke(this, new CrossRoadControllerEventArgs(State, Main, Secondary, Tram, Pedestrian));
+                                await Task.Delay(blinkTimer);
```
Blinking logic can be implemented as a TrafficLight class logic and not the controller logic. Some traffic lights can blick others cannot. It is not a responsibility of the controller to tell the traffic light how to blink.

ValeriiB commented on CrossRoadControllerShowModule.cs:
```diff
@@ -0,0 +1,373 @@
+                Console.ResetColor();
+                Console.WriteLine("|");
+                Console.ResetColor();
+                Console.WriteLine("---");
```
TrafficLightі should not know how to present itself on a UI. In WinForms application thare will be no Console. Let TrafficLightPresenter to decide how to present the TrafficLight of the UI.

Vlad-Slyzkoukhyi committed:
Re-worked implementation with usind set of data for traffic lighters states and time.
Presentation menu was fully moved to presentation side.
Totally optimazed all clases.

Vlad-Slyzkoukhyi committed:
Re-worked implementation with usind set of data for traffic lighters states and time.
Presentation menu was fully moved to presentation side.
Totally optimazed all clases.

Vlad-Slyzkoukhyi committed:
Re-worked implementation with usind set of data for traffic lighters states and time.
Presentation menu was fully moved to presentation side.
Totally optimazed all clases.

Vlad-Slyzkoukhyi committed:
Re-worked implementation with usind set of data for traffic lighters states and time.
Presentation menu was fully moved to presentation side.
Totally optimazed all clases.

Vlad-Slyzkoukhyi committed:
Re-worked implementation with usind set of data for traffic lighters states and time.
Presentation menu was fully moved to presentation side.
Totally optimazed all clases.

ValeriiB commented on CrossRoadController.cs:
```diff
@@ -0,0 +1,195 @@
+                            case TrafficLightsNames.PedestrianTL:
+                                PedestrianTrafficLighterEvent?.Invoke(new PedestrianTrafficLighterEventArgs(trafficLightState.PedestrianTLState, trafficLighter.Value, time));
+                                break;
+                        }
```
What if tomorrow I want to add one more traffic lights to the crossroad, for example TrainTrafficLight? In that case I think that the logic of the controller should be changed, because you tightly coupled your code with the types of the traffic lights that you have in the system. Look again in the pseudocode that I propose. I my case the controller knows nothing about the traffic lights that we have in the system. The controller just controls states of any type of a traffic lights we have in the syste. It say "hey, TrafficLights (and I do not care what type you are), change you state and stay in that state X milliseconds":
```
DO
GET next state (circularly) from states in the day mode (you defined 10 states on you diagram)
SET crossroad state
WAIT
WHILE(DayMode)
```
Make you controller to control states of any traffic lights in the system.

ValeriiB commented on CrossRoadController.cs:
```diff
@@ -0,0 +1,195 @@
+        }
+        internal static Dictionary<CrossRoadsStates, CrossRoadController> BuildDayModeCollection()
+        {
+            return new Dictionary<CrossRoadsStates, CrossRoadController>
```
Are you creating a collection of CrossRoadControllers? The name of the controller if "CrossRoad..." what statess that the controller controles a crossroad. How many controllers do you need to control ONE crossroad?

Think again what does the controller controls (or read my comment from the previos pull request?
It controls states of the CROSSROAD.  Each state of the crossroad consists of the traffic lights states that we have in the system.
And here you describe not a controller but one of a states of the crossroad 
```
new()
{
    MainTLState = StatesCondition.RED,
    SecondaryTLState = StatesCondition.RED,
    TramTLState = StatesCondition.STOP,
    PedestrianTLState = StatesCondition.RED,
    Time = 0 
}
```

A mistate that you did in you design is that you combined the "Controller" and "CrossRoadState" in one class.
Separate those two entoties.

ValeriiB commented on CrossRoadController.cs:
```diff
@@ -0,0 +1,195 @@
+            TramTL = 2,
+            PedestrianTL = 3,
+        }
+        internal Dictionary<int, TrafficLightsNames> TrafficLightersNames = new()
```
Why do you need this dictionary? What are the difference between the key and the value in it?
And the answer is: no difference because, for example, the value of "TrafficLightsNames.MainTL" is equal to 0 and it is absolutely the same as the coresponding key in the dictionary.

ValeriiB commented on PedestriamTrafficLighterShowModule.cs:
```diff
@@ -4,21 +4,62 @@
+        internal bool GreenLamp { get; set; }
+        internal static int blinkDuratation = 500;
+        internal static void ShowPedestrianTrafficLight(PedestrianTrafficLighterEventArgs e)
         {
```
The name of the class "PedestrianTrafficLighterEventArgs" states that it is a collection of some arguments.
Why the collection of arguments is responsible to "show" something? In other words how "ShowPedestrianTrafficLight" ir related to the collection of arguments?

ValeriiB commented on PedestriamTrafficLighterShowModule.cs:
```diff
@@ -4,21 +4,62 @@
+                    e.RedLamp = false;
+                    e.GreenLamp = false;
+                    break;
+            }
```
Above you set a state of the traffic lights.
And below you show that state in the console.
Reread my comment from the previous pull request:
> TrafficLightі should not know how to present itself on a UI. In WinForms application thare will be no Console. Let TrafficLightPresenter to decide how to present the TrafficLight of the UI.

ValeriiB replied:
Again you mistake is that you combined two entities "TrafficLight" and "Presenter" in one class. And in addition to that you gave the name to the class that is related neither to "TrafficLight" nor "Presenter": PedestrianTrafficLighterEventArgs.

To understand how to implement the system describe it in human words:

We have _**trafficlights**_ on _**crossroad**_ that traffic lights are working together, so there should be some _**controller**_ that controls the situation (or state) on that crossroad. The _**state of the crossroad**_ defines _**state of each traffic light**_". _**Lamps**_ of a specific traffic light shows a state of that traffic light.

I highlited nouns in the text and those nowns are potential names of the classes that you need in the system. Ofcouse some of them need better names, for example instead of "Lamps" you can use "TrafficLightPresenter" or "TrafficLightView" but the meaning of the term should be the same.
